sampler_remote.go

```diff
@@ -134,71 +159,164 @@ func (s *RemotelyControlledSampler) setSampler(sampler SamplerV2) {
+				if err := ps.Update(probabilistic.SamplingRate); err != nil {
+					return nil, err
+				}
+				return sampler, nil
 			}
-		} else {
-			s.sampler = newProbabilisticSampler(probabilistic.SamplingRate)
+			return newProbabilisticSampler(probabilistic.SamplingRate), nil
 		}
-	} else if rateLimiting := res.GetRateLimitingSampling(); rateLimiting != nil {
-		if rl, ok := s.sampler.(*RateLimitingSampler); ok {
-			rl.Update(float64(rateLimiting.MaxTracesPerSecond))
-		} else {
-			s.sampler = NewRateLimitingSampler(float64(rateLimiting.MaxTracesPerSecond))
+	}
+	return nil, nil
+}
+
+// -----------------------
+
+// RateLimitingSamplerUpdater is used by RemotelyControlledSampler to parse sampling configuration.
+type RateLimitingSamplerUpdater struct{}
+
+func (u *RateLimitingSamplerUpdater) Update(sampler SamplerV2, strategy interface{}) (SamplerV2, error) {
+	type response interface {
+		GetRateLimitingSampling() *sampling.RateLimitingSamplingStrategy
+	}
+	var _ response = new(sampling.SamplingStrategyResponse) // sanity signature check
+	if resp, ok := strategy.(response); ok {
+		if rateLimiting := resp.GetRateLimitingSampling(); rateLimiting != nil {
+			rateLimit := float64(rateLimiting.MaxTracesPerSecond)
+			if rl, ok := sampler.(*RateLimitingSampler); ok {
+				rl.Update(rateLimit)
+				return rl, nil
+			}
+			return NewRateLimitingSampler(rateLimit), nil
 		}
-	} else {
-		return fmt.Errorf("unsupported sampling strategy type %v", res.GetStrategyType())
 	}
-	return nil
+	return nil, nil
 }
 
-type httpSamplingManager struct {
+// -----------------------
+
+// AdaptiveSamplerUpdater is used by RemotelyControlledSampler to parse sampling configuration.
+type AdaptiveSamplerUpdater struct {
+	MaxOperations int // required
+}
+
+func (u *AdaptiveSamplerUpdater) Update(sampler SamplerV2, strategy interface{}) (SamplerV2, error) {
+	type response interface {
+		GetOperationSampling() *sampling.PerOperationSamplingStrategies
+	}
+	var _ response = new(sampling.SamplingStrategyResponse) // sanity signature check
+	if p, ok := strategy.(response); ok {
+		if operations := p.GetOperationSampling(); operations != nil {
+			if as, ok := sampler.(*AdaptiveSampler); ok {
+				as.update(operations)
+				return as, nil
+			}
+			return newAdaptiveSampler(operations, u.MaxOperations), nil
+		}
+	}
+	return nil, nil
+}
+
+// -----------------------
+
+type httpSamplingStrategyFetcher struct {
 	serverURL string
+	logger    Logger
 }
 
-func (s *httpSamplingManager) GetSamplingStrategy(serviceName string) (*sampling.SamplingStrategyResponse, error) {
-	var out sampling.SamplingStrategyResponse
+func (f *httpSamplingStrategyFetcher) Fetch(serviceName string) ([]byte, error) {
 	v := url.Values{}
 	v.Set("service", serviceName)
-	if err := utils.GetJSON(s.serverURL+"?"+v.Encode(), &out); err != nil {
+	uri := f.serverURL + "?" + v.Encode()
+
+	// TODO create and reuse http.Client with proper timeout settings, etc.
+	resp, err := http.Get(uri)
```

Comment:
I think we should do this soon - IIRC the default client doesn't specific things like timeouts/etc.

Reply:
we could, but not in this PR.